Add software versions chapter (#19)
fhdsl/Tools_for_Reproducible_Workflows_in_R: 08-software-versions.Rmd

Reviewer: Looks great @cansavvy! I wonder if we could talk a bit more about use cases of sharing and building a renv snapshot? like future self or someone else...maybe just adding to the slide about this? like an older version of the parrot or just one of your other characters from previous courses?

Reviewer: I think this comment was from before...

Author: Added a couple sentences about use cases for renv.

ottrpal::set_knitr_image_path()
```

## Learning Objectives

> ```{r, fig.align='center', out.width="100%", echo = FALSE, fig.alt= "In this software versions chapter our learning objectives are to be able to: Recognize that software versions influence data analysis results and reproducibility. Record packages used for an analysis. Use renv to make an R environment shareable to collaborators"}

Reviewer: could add another learning objective about recognizing other methods

Author: Adding `Recognize containerization as a method to share your entire computing environment with others`

## Software and package versions affect results!

Sometimes if we have generally the same software installed for reproducing an analysis, we may feel that that is "close enough". And given all the other technical aspects of reproducibility, it can be easy to overlook what versions of software packages we are using. However, controlling for software versions is critical for creating reproducible analyses. Software versions can directly affect not only whether an analysis will be able to run, but the results of the analysis [@BeaulieuJones2017].

## Session Info

Perhaps the easiest way to begin to address computing environment variability is to record what the computing environment looks like at the time an analysis is run. In R, this is a fairly straightforward task.

Generally at the end of your R notebook, you will want to print out your session info. You can do this by running the function `sessionInfo()` or the tidyverse version of this function from the devtools package, `devtools::session_info()`.

> We can run sessionInfo in this book (this book is powered by R).

Reviewer: ```suggestion
We can run `sessionInfo` in this book (this book was created using R tools).
```

Author: Is this redundant with my `(this book is powered by R)` or do you prefer this phrasing?

Reviewer: oops yes - but maybe created using R explains more than powered by? for people who are unfamiliar with that phrasing?

Reviewer: I edited it- is that better?

Author: Yup! I'll use that.

```{r}
sessionInfo()
```

Now we have recorded what some key aspects of our computing environment looked like at the time that this book was rendered last.
This print out may seem like a lot of nonsense at first, but it gives us some useful information in a pinch!

```{r, fig.align='center', out.width="100%", echo = FALSE, fig.alt= "Reproducible parrot is confused because their results are different after they have re-run their analysis. The parrot says: ‘Hmm… why am I getting a different result this time? Good thing I can check the session info to see if package versions might have caused this change!’"}
ottrpal::include_slide("https://docs.google.com/presentation/d/1MNHf8JpolaEP_vQ_kB-1xRBF9wo3haCArRu117hBoHA/edit#slide=id.g201bd406763_37_1030")
```

> If we take a look at two different session info print outs, we can begin to spot the differences. These differences may give us clues into why an analysis ran differently.

Reviewer: ```suggestion
If we take a look at two different session info printouts, we can begin to spot the differences. These differences may give us clues into why an analysis ran differently.
```

> ```{r, fig.align='center', out.width="100%", echo = FALSE, fig.alt= "Two session info print outs are show side by side.Highlighted we can see that they have different R versions: 4.0.2 vs 4.0.5. They also have different operating systems. The packages they have attached is rmarkdown but they also have different rmarkdown package versions!  If there are  discrepancies in re-runs of the analysis, the session info print out gives a record which may have clues to why that might be! This can give items to look into for determining why the results didn’t reproduce as expected."}

Reviewer: ```suggestion
```{r, fig.align='center', out.width="100%", echo = FALSE, fig.alt= "Two session info printouts are show side by side.Highlighted we can see that they have different R versions: 4.0.2 vs 4.0.5. They also have different operating systems. The packages they have attached is rmarkdown but they also have different rmarkdown package versions!  If there are  discrepancies in re-runs of the analysis, the session info printout gives a record which may have clues to why that might be! This can give items to look into for determining why the results didn’t reproduce as expected."}
```

Reviewer: OK there are more of these bellow... I can see why maybe you want to do "session info" instead but printout is a word so might just want to do that?

Reviewer: I will edit away my sessionInfo edits

For that, we need a slightly more involved solution of using [`renv`](https://rstudio.github.io/renv/articles/renv.html). `renv` is an R package that allows you to take 'snapshots' of your R computing environment and use those to track, share, and build R environments.

```{r, fig.align='center', out.width="100%", echo = FALSE, fig.alt= "In general, package managers work by capturing a snapshot of the environment and when that environment snapshot is shared, it attempt to rebuild it. In this example we show one computing environment, and using a package manager, we can take  snapshot of it. That snapshot can be shared to another computer which can be used to attempt to build the computing environment on this computer. This will help address some differences in package versions between two individual’s computers. "}
ottrpal::include_slide("https://docs.google.com/presentation/d/1MNHf8JpolaEP_vQ_kB-1xRBF9wo3haCArRu117hBoHA/edit#slide=id.g201bd406763_37_1492")
```

The `renv` workflow looks like this (as described by their documentation):

> > 1. Call `renv::init()` to initialize a new project-local environment with a private R library,

Reviewer: I know these are quotes but maybe don't need the comma or period at the ends of these since they are inconsistent

> > 1. Call `renv::init()` to initialize a new project-local environment with a private R library,

Reviewer: ```suggestion
> 1. Call `renv::init()` to initialize a new project-local environment with a private R library

> > 2. Work in the project as normal, installing and removing new R packages as they are needed in the project,

Reviewer: ```suggestion
> 2. Work in the project as normal, installing and removing new R packages as they are needed in the project
```

> > 3. Call `renv::snapshot()` to save the state of the project library to the lockfile (called renv.lock),

Reviewer: ```suggestion
> 3. Call `renv::snapshot()` to save the state of the project library to the lockfile (called renv.lock)
```

> > 4. Continue working on your project, installing and updating R packages as needed.

Reviewer: ```suggestion
> 4. Continue working on your project, installing and updating R packages as needed
```

> > 5. Call `renv::snapshot()` again to save the state of your project library if your attempts to update R packages were successful, or call `renv::restore()` to revert to the previous state as encoded in the lockfile if your attempts to update packages introduced some new problems.

Reviewer: ```suggestion
> 5. Call `renv::snapshot()` again to save the state of your project library if your attempts to update R packages were successful, or call `renv::restore()` to revert to the previous state as encoded in the lockfile if your attempts to update packages introduced some new problems
```

1. Be sure to commit and push the `renv.lock` file to your GitHub repository for your project.
2. Be sure to describe that your project uses `renv` in the README of this project (commit and push this to your GitHub repository also).

```{r, fig.align='center', out.width="100%", echo = FALSE, fig.alt= "The renv workflow begins with intializing an environment with renv::init(). Then you install or remove packages and otherwise work in R as normal. When you are ready to update the renv snapshot, you run renv::snapshot(). You can now share this environment snapshot on GitHub or wherever. The environment can be restored using renv::restore()"}
ottrpal::include_slide("https://docs.google.com/presentation/d/1MNHf8JpolaEP_vQ_kB-1xRBF9wo3haCArRu117hBoHA/edit#slide=id.g21dfe2f76f9_90_163")
```

The limitations of this method, [as noted by the `renv` authors](https://rstudio.github.io/renv/articles/renv.html#caveats), is that it really only tracks packages in R and cannot help track or enforce items that may affect the computing environment outside of R. So while it will aid in the reproducibility of your analysis, it will not cover everything.

## Containerization

> In order to truly reproduce a result with an identical computing environment you would need to use a containerized approach. To containerize a computing environment is to truly create an environment that is shippable to others. A container is analogous to a virtual machine. A computer runs a computing environment inside of it that is separate from the rest of the computer (hence why its called a container).

Reviewer: maybe define virtual machine too
maybe make a dictionary box about containerization (possible add virtual machine too it too)

Reviewer: could also add Docker and Singularity to that

Author: Added a dictionary box!

One of the most popular containerization softwares is Docker. Docker allows you to build your computing environment and share it on its online platform in the form of images that you can download and run. In fact, this book is rendered by a Docker container!

We will not cover Docker here but if you are interested in using a containerized approach like Docker, here are additional resources for learning:

- [Software Carpentries course on Docker](https://carpentries-incubator.github.io/docker-introduction/introduction/index.html)
- [ITCR Training Network chapters about Docker](https://jhudatascience.org/Adv_Reproducibility_in_Cancer_Informatics/launching-a-docker-image.html)

> - [Docker documentation about getting started](https://www.docker.com/get-started/)

Reviewer: maybe we could also link to singularity? and say that they can be used together for PHI? (apparently that is the case... maybe we can find a link for that)

Author: There's nothing particularly bad about Docker. People just sometimes do bad things like trying to store data on the docker image itself which is the problem. Singularity apparently has some encryption though that can make it more secure if someone were to do this. I've added some sources for both Singularity and how to use Docker properly.

Author: Done! Added info about Singularity!

## Conclusion

In summary:

- Software versions affect the reproducibility of an analysis
- Printing out session info is a great way to record software versions
- `renv` is an R package that allows you to share your R specific computing environment

> - Containerization softwares like Docker allow you to more completely share a replicate computing environment.

Reviewer: ```suggestion
- Containerization softwares like Docker allow you to more completely share a replicate computing environment
```
